### ISSUE TITLE
No IUserTokenProvider is registered.

### ISSUE THREAD

**Issue by emaish** [aspnetboilerplate/module-zero#50]
when i try to use await GeneratePasswordResetTokenAsync(adminUser.Id);
i got exception No IUserTokenProvider is registered, i searched for that error and they said that UserTokenProvider isn't resolved while resolving UserManager, and they suggest solution by this code
```C#
var provider = new DpapiDataProtectionProvider("YourAppName");
    UserManager.UserTokenProvider = new DataProtectorTokenProvider<User, string>(provider.Create("UserToken")) 
        as IUserTokenProvider<User, string>;
```
where to put it and maintain the dependency ?

**Comment by hikalkan** [comment]
Yes, I got the same issue yesterday while implementing password reset with email feature.
I'll implement it.
But you can do it yourself using User.PasswordResetCode.

**Comment by ahaverdings** [comment]
When using the DataProtectorTokenProvider the User.PasswordResetCode StringLength is too short to store the token in the DB.

**Comment by hikalkan** [comment]
Thanks, I'll make it bigger.

**Comment by hikalkan** [comment]
I made it 128. Hope it will be enough.

**Comment by lerich** [comment]
it's return StringLength 328

**Comment by hikalkan** [comment]
OK, I'll check and increment it if needed.

**Comment by daniel-chenery** [comment]
Will implementing this fix the problem I'm running into on Azure? (And presumably other shared hosting environments)
http://forum.aspnetboilerplate.com/viewtopic.php?f=3&t=386

**Comment by hikalkan** [comment]
I've no idea. Has anybody tried it?

**Comment by daniel-chenery** [comment]
I'm implementing it in my UserManager like so:

```c#
var provider = new DpapiDataProtectionProvider("MyApp");
this.UserTokenProvider = new DataProtectorTokenProvider<User, long>(provider.Create("UserToken"));
```

Running into the same problem though, something to do with user MachineKeys over SessionKeys

**Comment by Znow** [comment]
When is this being fixed? I cant use the .ResetPassword function due to the lack of IUserTokenProvider.

**Comment by daniel-chenery** [comment]
@Znow If you use my snippet above it should get you somewhere :)

**Comment by Znow** [comment]
@daniel-chenery Thanks, got me a bit further. But .ResetPassword() and .ResetPasswordAsync methods, doesn't reset the password at all, all though no errors are thrown. It doesn't get updated in the DB...

**Comment by daniel-chenery** [comment]
@Znow It's been a while, but I think you need to it manually
Something like
```
var myUser = _userRepo.Get(1);
var ResetCode = _userManager.GeneratePasswordResetToken()
var TrimmedCode = ResetCode.Substring(0, User.MaxPasswordResetCode)

myUser.PasswordResetCode = TrimmedCode
```

**Comment by Znow** [comment]
@daniel-chenery I got that part done, nothing there. But the actual method _.ResetPassword()_ on UserManager, does not work. It does not update the password in the database. Nothing to do with the password reset code, already got that part working fine :-)

**Comment by daniel-chenery** [comment]
@Znow Ah, I see. This issue was relating to the token provider.
Can you not just do it manually?

It should be something along the lines of...
```c#
if(myUser.PasswordResetCode == input.PasswordResetCode) {
myUser.Password = _userManager.PasswordHasher.HashPassword(model.Password)
}
```

**Comment by Znow** [comment]
@daniel-chenery I could indeed, good point, but why are there methods for the sake of this, when they are not functional? :P

Thanks so far :-)

**Comment by emaish** [comment]
I successfully implemented it, and tested it on Microsoft Azure, maybe @hikalkan could suggest enhancement.

- GenerateEmailConfirmationTokenAsync is working
- VerifyUserTokenAsync is working

Now let's see my implementation

**Changes needed in Owin Startup Class**
1- define static IDataProtectionProvider
```C#
internal static IDataProtectionProvider dataProtectionProvider { get; private set; }
```
2- Set dataProtectionProvider in the Configuration method
```C#
public void Configuration(IAppBuilder app)
{
    dataProtectionProvider = app.GetDataProtectionProvider();
    app.UseAbp();
    // other configurations
    // ....
}
```
**Changes needed in UserManager**
Changes in UserManager require Owin Startup access and also require reference to Microsoft.AspNet.Identity.Owin which are not available in Core Project, so i created CustomUserManager in Web Project which inherit from UserManager in Core Project, and i use this CustomUserManager in Web Project Code.

1- Put the following code in CutstomUserManager constructor
```C#
var dataProtectionProvider = Startup.dataProtectionProvider;
this.UserTokenProvider =
        new DataProtectorTokenProvider<User, long>(dataProtectionProvider.Create("ASP.NET Identity"));
```
Maybe @hikalkan find a solution to avoid creating CustomUserManager in Web Project

**Comment by hikalkan** [comment]
Hi,

Based on @emaish's solution (and after investigating source of of Microsoft Identity), I a little bit simplified it.

Add this to OWIN Startup class (just after app.UseAbp()):

````C#
app.RegisterDataProtectionProvider();
````

That's all.

**Comment by gabrielrobert** [comment]
@hikalkan I'm still having the issue using AspNetCore, is there a way to handle this?

```
System.NotSupportedException: No IUserTokenProvider is registered.
   at Microsoft.AspNet.Identity.UserManager`2.<GenerateUserTokenAsync>d__fe.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at TestProject.Web.Controllers.AccountController.<Register>d__13.MoveNext() in C:\dev\TestProject\src\TestProject.Web\Controllers\AccountController.cs:line 175
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionFilterAsync>d__28.MoveNext()
```

**Comment by hikalkan** [comment]
Can you create a nother issue for that (since we closed this in v1.0). We will try it and inform you.
